In [183]:
import pandas as pd
link='https://github.com/AdriMA3/PC3_1/raw/refs/heads/main/Reserves%20of%20foreign%20exchange%20and%20gold%20(3).csv'
reserves=pd.read_csv(link)
link='https://github.com/AdriMA3/PC3_1/raw/refs/heads/main/Energy%20consumption%20per%20capita.csv'
energy=pd.read_csv(link)
link='https://github.com/AdriMA3/PC3_1/raw/refs/heads/main/Electricity%20-%20installed%20generating%20capacity.csv'
electricity=pd.read_csv(link)
link='https://github.com/AdriMA3/PC3_1/raw/refs/heads/main/Education%20expenditures.csv'
education=pd.read_csv(link)

In [184]:
reserves

name                   slug               value  \
0                    China                  china  $3,265,000,000,000   
1                    Japan                  japan  $1,160,000,000,000   
2              Switzerland            switzerland    $822,130,000,000   
3                   Russia                 russia    $597,217,000,000   
4                    India                  india    $569,544,000,000   
..                     ...                    ...                 ...   
190            South Sudan            south-sudan         $72,881,000   
191           Burkina Faso           burkina-faso         $49,000,000   
192             Montserrat             montserrat         $47,580,000   
193  Sao Tome and Principe  sao-tome-and-principe         $46,247,000   
194                Somalia                somalia         $30,450,000   

     date_of_information  ranking                             region  
0                   2024        1            East and Southeast Asia  
1                   2024        2            East and Southeast Asia  
2                   2024        3                             Europe  
3                   2023        4                       Central Asia  
4                   2024        5                         South Asia  
..                   ...      ...                                ...  
190                 2023      191                             Africa  
191                 2017      192                             Africa  
192                 2017      193  Central America and the Caribbean  
193                 2023      194                             Africa  
194                 2014      195                             Africa  

[195 rows x 6 columns]

In [185]:
energy

name                              slug  \
0                                Qatar                             qatar   
1                            Singapore                         singapore   
2                              Bahrain                           bahrain   
3                 United Arab Emirates              united-arab-emirates   
4                               Brunei                            brunei   
..                                 ...                               ...   
190                               Chad                              chad   
191  Congo, Democratic Republic of the  congo-democratic-republic-of-the   
192           Central African Republic          central-african-republic   
193                            Burundi                           burundi   
194                            Somalia                           somalia   

      Btu/person  date_of_information  ranking                   region  
0    814,308,000                 2023        1              Middle East  
1    643,259,000                 2023        2  East and Southeast Asia  
2    554,202,000                 2023        3              Middle East  
3    450,432,000                 2023        4              Middle East  
4    403,365,000                 2023        5  East and Southeast Asia  
..           ...                  ...      ...                      ...  
190    1,502,000                 2023      191                   Africa  
191    1,305,000                 2023      192                   Africa  
192      954,000                 2023      193                   Africa  
193      946,000                 2023      194                   Africa  
194      649,000                 2023      195                   Africa  

[195 rows x 6 columns]

In [186]:
electricity

name  \
0                                            China   
1                                    United States   
2                                            India   
3                                            Japan   
4                                           Russia   
..                                             ...   
206                                       Kiribati   
207              Falkland Islands (Islas Malvinas)   
208                                     Montserrat   
209  Saint Helena, Ascension, and Tristan da Cunha   
210                                           Niue   

                                            slug             kW  \
0                                          china  2,949,000,000   
1                                  united-states  1,235,000,000   
2                                          india    499,136,000   
3                                          japan    361,617,000   
4                                         russia    301,926,000   
..                                           ...            ...   
206                                     kiribati         12,000   
207              falkland-islands-islas-malvinas         10,000   
208                                   montserrat          6,000   
209  saint-helena-ascension-and-tristan-da-cunha          5,000   
210                                         niue          3,000   

     date_of_information  ranking                             region  
0                   2023        1            East and Southeast Asia  
1                   2023        2                      North America  
2                   2023        3                         South Asia  
3                   2023        4            East and Southeast Asia  
4                   2023        5                       Central Asia  
..                   ...      ...                                ...  
206                 2023      207              Australia and Oceania  
207                 2023      208                      South America  
208                 2023      209  Central America and the Caribbean  
209                 2023      210                             Africa  
210                 2023      211              Australia and Oceania  

[211 rows x 6 columns]

In [187]:
education

name              slug  % of GDP  date_of_information  \
0    Marshall Islands  marshall-islands      13.6                 2020   
1     Solomon Islands   solomon-islands      12.8                 2020   
2            Kiribati          kiribati      12.4                 2019   
3           Greenland         greenland      10.2                 2019   
4             Bolivia           bolivia       9.8                 2020   
..                ...               ...       ...                  ...   
192  Papua New Guinea  papua-new-guinea       1.4                 2020   
193         Venezuela         venezuela       1.3                 2017   
194            Monaco            monaco       1.2                 2019   
195           Nigeria           nigeria       0.5                 2013   
196           Somalia           somalia       0.3                 2019   

     ranking                   region  
0          1    Australia and Oceania  
1          2    Australia and Oceania  
2          3    Australia and Oceania  
3          4            North America  
4          5            South America  
..       ...                      ...  
192      193  East and Southeast Asia  
193      194            South America  
194      195                   Europe  
195      196                   Africa  
196      197                   Africa  

[197 rows x 6 columns]

Se elimina las columnas que no nos ayudaran para los datos que queremos analizar

In [188]:
reserves=reserves.drop(columns=['slug', 'ranking'])
energy=energy.drop(columns=['slug','ranking'])
electricity=electricity.drop(columns=['slug','ranking'])
education=education.drop(columns=['slug','ranking'])

In [189]:
reserves

name               value  date_of_information  \
0                    China  $3,265,000,000,000                 2024   
1                    Japan  $1,160,000,000,000                 2024   
2              Switzerland    $822,130,000,000                 2024   
3                   Russia    $597,217,000,000                 2023   
4                    India    $569,544,000,000                 2024   
..                     ...                 ...                  ...   
190            South Sudan         $72,881,000                 2023   
191           Burkina Faso         $49,000,000                 2017   
192             Montserrat         $47,580,000                 2017   
193  Sao Tome and Principe         $46,247,000                 2023   
194                Somalia         $30,450,000                 2014   

                                region  
0              East and Southeast Asia  
1              East and Southeast Asia  
2                               Europe  
3                         Central Asia  
4                           South Asia  
..                                 ...  
190                             Africa  
191                             Africa  
192  Central America and the Caribbean  
193                             Africa  
194                             Africa  

[195 rows x 4 columns]

Cambiamos el nombre de las columnas para un mejor analisis

In [190]:
#cambiaremos el nombre de las columnas
reserves=reserves.rename(columns={'name':'country','value':'reserves_US','date_of_information':'reserves_yearData','region':'region'})
energy=energy.rename(columns={'name':'country',' Btu/person':'Btu_energy','date_of_information':'energy_yearData','region':'region'})
electricity=electricity.rename(columns={'name':'country',' kW':'kW_electricity','date_of_information':'electricity_yearData','region':'region'})
education=education.rename(columns={'name':'country','% of GDP':'%_PBI_education','date_of_information':'education_yearData','region':'region'})

In [191]:
reserves

country         reserves_US  reserves_yearData  \
0                    China  $3,265,000,000,000               2024   
1                    Japan  $1,160,000,000,000               2024   
2              Switzerland    $822,130,000,000               2024   
3                   Russia    $597,217,000,000               2023   
4                    India    $569,544,000,000               2024   
..                     ...                 ...                ...   
190            South Sudan         $72,881,000               2023   
191           Burkina Faso         $49,000,000               2017   
192             Montserrat         $47,580,000               2017   
193  Sao Tome and Principe         $46,247,000               2023   
194                Somalia         $30,450,000               2014   

                                region  
0              East and Southeast Asia  
1              East and Southeast Asia  
2                               Europe  
3                         Central Asia  
4                           South Asia  
..                                 ...  
190                             Africa  
191                             Africa  
192  Central America and the Caribbean  
193                             Africa  
194                             Africa  

[195 rows x 4 columns]

Con esto se ve que si cambiaron los nombres de las columnas,de igual manera sucedio con las otras tablas.

Ahora nos fijamos que debemos de eliminar ese signo de dolar

In [192]:
def clean_numeric(reserves, reserves_US):
    # Elimina el signo de dólar y las comas, y convierte a número
    reserves[reserves_US] = reserves[reserves_US].replace({'\$': '', ',': ''}, regex=True)
    reserves[reserves_US] = pd.to_numeric(reserves[reserves_US], errors='coerce')
    return reserves

In [193]:
reserves = clean_numeric(reserves, "reserves_US")

Ahora nos fijamos que tipo de dato es en cada columnas de cada tabla y cambiamos de tipo de dato si es necesario.

In [194]:
energy.dtypes

country            object
Btu_energy         object
energy_yearData     int64
region             object
dtype: object

Se cambia usando este codigo , y de igual manera se hara con la tabla de electricity

In [195]:
energy['Btu_energy']=energy['Btu_energy'].str.replace(',', '').astype(int)

In [196]:
electricity.dtypes

country                 object
kW_electricity          object
electricity_yearData     int64
region                  object
dtype: object

In [197]:
electricity['kW_electricity']=electricity['kW_electricity'].str.replace(',', '').astype(int)

Los tipos de datos int , float64 y object ya son corectos para cada columna de cada tabla.

Ahora veremos los paises que estan en todas la tablas y luego ver si hay algunos que estan mal escritos

In [198]:
print("reserves:", reserves['country'].nunique())
print("energy:", energy['country'].nunique())
print("electricity:", electricity['country'].nunique())
print("education:", education['country'].nunique())

reserves: 195
energy: 195
electricity: 211
education: 197


In [199]:
set1 = set(reserves['country'])
set2 = set(energy['country'])
set3 = set(electricity['country'])
set4 = set(education['country'])

# Intersección: países que están en las 4 tablas
common = set1 & set2 & set3 & set4
print("Países comunes en las 4 tablas:", len(common))

Países comunes en las 4 tablas: 176


Ahora veremos si los paises que no estan en la lista comun estan mal escritos y se puede modificar para tener más datos

In [200]:
# países en reserves, energy y education, pero NO en electricity
missing_in_electricity = (set1 & set2 & set4) - set3

missing_in_reserves = (set2 & set3 & set4) - set1
missing_in_energy = (set1 & set3& set4) - set2
missing_in_education = (set1& set2& set3) - set4

print("Posibles errores en electricity:", missing_in_electricity)
print("Posibles errores en reserves:", missing_in_reserves)
print("Posibles errores en energy:", missing_in_energy)
print("Posibles errores en education:", missing_in_education)

Posibles errores en electricity: set()
Posibles errores en reserves: {'Bermuda', 'British Virgin Islands', 'Puerto Rico', 'Kiribati'}
Posibles errores en energy: {'Montserrat', 'Iceland', 'Trinidad and Tobago'}
Posibles errores en education: {'Cuba', 'Syria', 'Equatorial Guinea', 'Bosnia and Herzegovina', 'Kosovo', 'Libya', 'Sudan', 'Eritrea', 'Montenegro', 'Yemen', 'North Macedonia', 'Taiwan'}


Ahora luego de examinar, se procede a unir los data frame.

In [202]:
for df in [reserves, energy, electricity, education]:
    df['country'] = df['country'].str.strip().str.lower()
    df['region'] = df['region'].str.strip().str.lower()

In [203]:
reserves_sel = reserves[['country', 'region', 'reserves_US', 'reserves_yearData']]
energy_sel = energy[['country', 'region', 'Btu_energy', 'energy_yearData']]
electricity_sel = electricity[['country', 'region', 'kW_electricity', 'electricity_yearData']]
education_sel = education[['country', 'region', '%_PBI_education', 'education_yearData']]

In [204]:
df_merged = reserves_sel.merge(energy_sel, on=['country', 'region'], how='outer')
df_merged = df_merged.merge(electricity_sel, on=['country', 'region'], how='outer')
df_merged = df_merged.merge(education_sel, on=['country', 'region'], how='outer')

In [205]:
df_merged

country                             region   reserves_US  \
0       afghanistan                         south asia  9.749000e+09   
1           albania                             europe  6.455000e+09   
2           algeria                             africa  6.844800e+10   
3    american samoa              australia and oceania           NaN   
4           andorra                             europe           NaN   
..              ...                                ...           ...   
215  virgin islands  central america and the caribbean           NaN   
216       west bank                        middle east  1.328000e+09   
217           yemen                        middle east  1.251000e+09   
218          zambia                             africa  3.173000e+09   
219        zimbabwe                             africa  2.613730e+08   

     reserves_yearData  Btu_energy  energy_yearData  kW_electricity  \
0               2020.0   3380000.0           2023.0        627000.0   
1               2023.0  27407000.0           2023.0       2857000.0   
2               2024.0  61843000.0           2023.0      22591000.0   
3                  NaN  89105000.0           2019.0         50000.0   
4                  NaN         NaN              NaN             NaN   
..                 ...         ...              ...             ...   
215                NaN         NaN              NaN        326000.0   
216             2024.0  14991000.0           2023.0        352000.0   
217             2022.0   2987000.0           2023.0       1790000.0   
218             2023.0   8265000.0           2023.0       3986000.0   
219             2024.0  10855000.0           2023.0       2491000.0   

     electricity_yearData  %_PBI_education  education_yearData  
0                  2023.0              2.9              2020.0  
1                  2023.0              3.1              2020.0  
2                  2023.0              7.0              2020.0  
3                  2023.0              NaN                 NaN  
4                     NaN              2.9              2021.0  
..                    ...              ...                 ...  
215                2023.0              NaN                 NaN  
216                2023.0              5.3              2018.0  
217                2023.0              NaN                 NaN  
218                2023.0              3.7              2020.0  
219                2023.0              3.9              2018.0  

[220 rows x 10 columns]

Ahora solo ponemos los 176 paises que hay en común en las 4 tablas.

In [206]:
df_merged = reserves_sel.merge(energy_sel, on=['country', 'region'], how='inner')
df_merged = df_merged.merge(electricity_sel, on=['country', 'region'], how='inner')
df_merged = df_merged.merge(education_sel, on=['country', 'region'], how='inner')

In [207]:
df_merged

country                   region    reserves_US  \
0                    china  east and southeast asia  3265000000000   
1                    japan  east and southeast asia  1160000000000   
2              switzerland                   europe   822130000000   
3                   russia             central asia   597217000000   
4                    india               south asia   569544000000   
..                     ...                      ...            ...   
171                   togo                   africa       77800000   
172            south sudan                   africa       72881000   
173           burkina faso                   africa       49000000   
174  sao tome and principe                   africa       46247000   
175                somalia                   africa       30450000   

     reserves_yearData  Btu_energy  energy_yearData  kW_electricity  \
0                 2024   113805000             2023      2949000000   
1                 2024   129504000             2023       361617000   
2                 2024    99578000             2023        26502000   
3                 2023   224858000             2023       301926000   
4                 2024    25179000             2023       499136000   
..                 ...         ...              ...             ...   
171               2017     4538000             2023          326000   
172               2023     2092000             2023          136000   
173               2017     3481000             2023          749000   
174               2023     8875000             2023           29000   
175               2014      649000             2023          156000   

     electricity_yearData  %_PBI_education  education_yearData  
0                    2023              3.6                2020  
1                    2023              3.4                2020  
2                    2023              5.2                2020  
3                    2023              3.7                2020  
4                    2023              4.5                2020  
..                    ...              ...                 ...  
171                  2023              4.0                2020  
172                  2023              1.5                2016  
173                  2023              5.5                2020  
174                  2023              5.0                2020  
175                  2023              0.3                2019  

[176 rows x 10 columns]

In [208]:
# Agrupar por región y calcular promedio, min y max para cada variable
stats_by_region = df_merged.groupby('region').agg({
    'reserves_US': ['mean', 'min', 'max'],
    'Btu_energy': ['mean', 'min', 'max'],
    'kW_electricity': ['mean', 'min', 'max'],
    '%_PBI_education': ['mean', 'min', 'max']
})
# Mostrar resultado
stats_by_region

reserves_US                               \
                                           mean           min            max   
region                                                                         
africa                             6.860587e+09      30450000    68448000000   
australia and oceania              1.147904e+10     396530000    54455000000   
central america and the caribbean  4.561870e+09     155971000    23834000000   
central asia                       7.555822e+10    3237000000   597217000000   
east and southeast asia            3.536142e+11     781995000  3265000000000   
europe                             6.415875e+10     921269000   822130000000   
middle east                        9.432479e+10    1328000000   436769000000   
north america                      1.890850e+11  117551000000   227760000000   
south america                      4.748692e+10      87100000   318857000000   
south asia                         7.907593e+10     673203000   569544000000   

                                     Btu_energy                       \
                                           mean       min        max   
region                                                                 
africa                             1.412284e+07    649000   98847000   
australia and oceania              6.221671e+07   5655000  223158000   
central america and the caribbean  5.587300e+07   3486000  153952000   
central asia                       1.039161e+08  16192000  261142000   
east and southeast asia            1.186874e+08   6825000  643259000   
europe                             1.131422e+08  27407000  234698000   
middle east                        2.406509e+08  14991000  814308000   
north america                      2.158707e+08  57539000  311599000   
south america                      4.641467e+07  25733000   78496000   
south asia                         2.350638e+07   3380000   64082000   

                                  kW_electricity                         \
                                            mean        min         max   
region                                                                    
africa                              4.619820e+06      29000    65989000   
australia and oceania               1.706100e+07      34000   108193000   
central america and the caribbean   1.580500e+06      41000     6581000   
central asia                        4.239578e+07    3944000   301926000   
east and southeast asia             2.108446e+08     277000  2949000000   
europe                              4.027509e+07     779000   275658000   
middle east                         3.381729e+07     352000   119620000   
north america                       5.008580e+08  105586000  1235000000   
south america                       3.550408e+07     259000   240251000   
south asia                          7.211612e+07     432000   499136000   

                                  %_PBI_education             
                                             mean  min   max  
region                                                        
africa                                   4.268000  0.3   9.6  
australia and oceania                    6.314286  2.2  12.8  
central america and the caribbean        4.555000  1.4   8.7  
central asia                             4.333333  2.8   6.2  
east and southeast asia                  3.633333  1.4   6.3  
europe                                   5.151429  3.1   7.2  
middle east                              4.528571  1.7   7.8  
north america                            5.200000  4.3   6.1  
south america                            4.808333  1.3   9.8  
south asia                               3.850000  1.9   7.0

Ahora realizamos las etiquetas para las nuevas columnas

In [209]:
# Definimos etiquetas
etiquetas = ['Muy bajo', 'Bajo', 'Medio', 'Alto', 'Muy alto']

# Creamos columnas categorizadas usando etiquetas
df_merged['reserves_level'] = pd.qcut(df_merged['reserves_US'], q=5, labels=etiquetas)
df_merged['energy_level'] = pd.qcut(df_merged['Btu_energy'], q=5, labels=etiquetas)
df_merged['electricity_level'] = pd.qcut(df_merged['kW_electricity'], q=5, labels=etiquetas)
df_merged['education_level'] = pd.qcut(df_merged['%_PBI_education'], q=5, labels=etiquetas)
cols_to_show=['country','region','reserves_US','Btu_energy','kW_electricity','%_PBI_education','reserves_level','energy_level','electricity_level','education_level']

In [210]:
df_merged[cols_to_show]

country                   region    reserves_US  \
0                    china  east and southeast asia  3265000000000   
1                    japan  east and southeast asia  1160000000000   
2              switzerland                   europe   822130000000   
3                   russia             central asia   597217000000   
4                    india               south asia   569544000000   
..                     ...                      ...            ...   
171                   togo                   africa       77800000   
172            south sudan                   africa       72881000   
173           burkina faso                   africa       49000000   
174  sao tome and principe                   africa       46247000   
175                somalia                   africa       30450000   

     Btu_energy  kW_electricity  %_PBI_education reserves_level energy_level  \
0     113805000      2949000000              3.6       Muy alto     Muy alto   
1     129504000       361617000              3.4       Muy alto     Muy alto   
2      99578000        26502000              5.2       Muy alto         Alto   
3     224858000       301926000              3.7       Muy alto     Muy alto   
4      25179000       499136000              4.5       Muy alto         Bajo   
..          ...             ...              ...            ...          ...   
171     4538000          326000              4.0       Muy bajo     Muy bajo   
172     2092000          136000              1.5       Muy bajo     Muy bajo   
173     3481000          749000              5.5       Muy bajo     Muy bajo   
174     8875000           29000              5.0       Muy bajo         Bajo   
175      649000          156000              0.3       Muy bajo     Muy bajo   

    electricity_level education_level  
0            Muy alto            Bajo  
1            Muy alto            Bajo  
2                Alto            Alto  
3            Muy alto            Bajo  
4            Muy alto           Medio  
..                ...             ...  
171          Muy bajo           Medio  
172          Muy bajo        Muy bajo  
173              Bajo            Alto  
174          Muy bajo            Alto  
175          Muy bajo        Muy bajo  

[176 rows x 10 columns]

Se muestra en la tabla anterior como ddeberia de quedar.
Luego se procede a guardar la tabla con lo requerido.

In [211]:
#solo las columnas finales que piden
df_final = df_merged[[
    'country', 'region',
    'reserves_US', 'Btu_energy', 'kW_electricity', '%_PBI_education',
    'reserves_level', 'energy_level', 'electricity_level', 'education_level'
]]

# Guardamos como archivo CSV asi como guardamos en el inicio para poder abrir la tabla
df_final.to_csv("datos_finales.csv", index=False)

df_final

country                   region    reserves_US  \
0                    china  east and southeast asia  3265000000000   
1                    japan  east and southeast asia  1160000000000   
2              switzerland                   europe   822130000000   
3                   russia             central asia   597217000000   
4                    india               south asia   569544000000   
..                     ...                      ...            ...   
171                   togo                   africa       77800000   
172            south sudan                   africa       72881000   
173           burkina faso                   africa       49000000   
174  sao tome and principe                   africa       46247000   
175                somalia                   africa       30450000   

     Btu_energy  kW_electricity  %_PBI_education reserves_level energy_level  \
0     113805000      2949000000              3.6       Muy alto     Muy alto   
1     129504000       361617000              3.4       Muy alto     Muy alto   
2      99578000        26502000              5.2       Muy alto         Alto   
3     224858000       301926000              3.7       Muy alto     Muy alto   
4      25179000       499136000              4.5       Muy alto         Bajo   
..          ...             ...              ...            ...          ...   
171     4538000          326000              4.0       Muy bajo     Muy bajo   
172     2092000          136000              1.5       Muy bajo     Muy bajo   
173     3481000          749000              5.5       Muy bajo     Muy bajo   
174     8875000           29000              5.0       Muy bajo         Bajo   
175      649000          156000              0.3       Muy bajo     Muy bajo   

    electricity_level education_level  
0            Muy alto            Bajo  
1            Muy alto            Bajo  
2                Alto            Alto  
3            Muy alto            Bajo  
4            Muy alto           Medio  
..                ...             ...  
171          Muy bajo           Medio  
172          Muy bajo        Muy bajo  
173              Bajo            Alto  
174          Muy bajo            Alto  
175          Muy bajo        Muy bajo  

[176 rows x 10 columns]